File: app/src/main/java/com/kira/kdownloader/engine/ExtractorOptions.java
package com.kira.kdownloader.engine;

import java.net.URI;
import java.util.Locale;

public final class ExtractorOptions {
    private ExtractorOptions() {}

    public static boolean isTikTokUrl(String sourceUrl) {
        return hostMatches(sourceUrl, "tiktok.com");
    }

    public static boolean isYouTubeUrl(String sourceUrl) {
        return hostMatches(sourceUrl, "youtube.com") || hostMatches(sourceUrl, "youtu.be");
    }

    public static boolean isInstagramUrl(String sourceUrl) {
        return hostMatches(sourceUrl, "instagram.com") || hostMatches(sourceUrl, "instagr.am");
    }

    public static boolean isFacebookUrl(String sourceUrl) {
        return hostMatches(sourceUrl, "facebook.com")
                || hostMatches(sourceUrl, "fb.com")
                || hostMatches(sourceUrl, "fb.watch");
    }

    private static boolean hostMatches(String sourceUrl, String expectedHost) {
        try {
            String host = new URI(sourceUrl.trim()).getHost();
            if (host == null) return false;
            host = host.toLowerCase(Locale.ROOT);
            return host.equals(expectedHost) || host.endsWith("." + expectedHost);
        } catch (Exception ignored) {
            return false;
        }
    }
}

File: app/src/main/java/com/kira/kdownloader/engine/ExtractorOptions.kt
package com.kira.kdownloader.engine

import java.net.URI

internal object ExtractorOptions {
    private const val TIKTOK_HOST = "tiktok.com"
    private const val YOUTUBE_HOST = "youtube.com"
    private val installIdPattern = Regex("\\d{19}")

    fun isTikTokUrl(sourceUrl: String): Boolean {
        return hostMatches(sourceUrl, TIKTOK_HOST)
    }

    fun isYouTubeUrl(sourceUrl: String): Boolean {
        return hostMatches(sourceUrl, YOUTUBE_HOST) || hostMatches(sourceUrl, "youtu.be")
    }

    fun tikTokAppInfo(installIds: List<String>): String {
        require(installIds.isNotEmpty()) { "At least one TikTok install ID is required" }
        require(installIds.all(installIdPattern::matches)) {
            "TikTok install IDs must contain exactly 19 digits"
        }
        return "tiktok:app_info=${installIds.joinToString(",")}" 
    }

    private fun hostMatches(sourceUrl: String, expectedHost: String): Boolean {
        val host = runCatching { URI(sourceUrl.trim()).host }
            .getOrNull()
            ?.lowercase()
            ?: return false
        return host == expectedHost || host.endsWith(".$expectedHost")
    }
}
